Products: Grouping design for product creation type sheet

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #11361, #11211
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR is the second part of the product type creation bottom sheet.

Specifically, it now groups the "Manual product creation" into three categories:

1. Standard
   - Physical product
   - Virtual product
   - Variable product
2. Subscription
   - Simple subscription
   - Variable subscription
3. Other
   - Grouped product
   - External product


## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

1. Go to Products tab,
2. Create new product,
3. Ensure that the bottom sheet demo matches the video below, and with the design video on https://github.com/woocommerce/woocommerce-ios/issues/11361#issue-2027620643
4. Additionally, test with a site that doesn't have subscriptions support. Ensure that the subscriptions category and its product types are not shown.

## Video
<!-- Include before and after images or gifs when appropriate. -->


https://github.com/woocommerce/woocommerce-ios/assets/266376/4f2340dd-781b-4a76-9de3-1d1d93b09fe1


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

hafizrahman committed:
Add product creation categories.

hafizrahman committed:
Update ManualProductTypeOptions to display product options by category.

hafizrahman committed:
Various design updates.

hafizrahman committed:
Localization format updates.

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Refactor the product categories order into a Constant instead since it's not likely to be reused.

hafizrahman committed:
More styling updates.

hafizrahman committed:
Refactor grouped product type to be excluded if there's no product types in a category.

This can happen if, for example, a store doesn't support subscription products creation.

hafizrahman committed:
Refactor placement for ProductCreationCategory

hafizrahman committed:
Update some product type labels to match design.

hafizrahman committed:
Update release notes.

hafizrahman commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/BottomSheetListSelector/BottomSheetProductType.swift:
```diff
@@ -77,13 +76,13 @@ public enum BottomSheetProductType: Hashable, Identifiable {
                 return NSLocalizedString(
-                    "bottomSheetProductType.simpleVirtual.title",
-                    value: "Simple virtual product",
+                    "bottomSheetProductType.simpleVirtualProduct.title",
```
Have to also change the key for the string value change to apply. Not completely sure why, but without doing this, the string change doesn't appear in the simulator.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Add Product/ManualProductTypeOptions.swift:
```diff
@@ -6,37 +6,70 @@ struct ManualProductTypeOptions: View {
+                    .padding(.bottom, Constants.categoryVerticalSpacing)
+                    .padding(.horizontal, Constants.horizontalSpacing)
+
+                ForEach(productTypes, id: \.self) { productType in
```
Nit: `BottomSheetProductType` is already conforming to `Identifiable`, so we can omit the `id` parameter here.

hafizrahman replied:
👍🏼  updated in https://github.com/woocommerce/woocommerce-ios/pull/12273/commits/34442d134d84058e62685dcfb68a6ecdfc1bae34

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Add Product/ManualProductTypeOptions.swift:
```diff
@@ -6,37 +6,70 @@ struct ManualProductTypeOptions: View {
-                        .subheadlineStyle()
+        ForEach(groupedProductTypes, id: \.0) { (category, productTypes) in
+            VStack {
+                Text(category.label)
+                    .subheadlineStyle()
+                    .textCase(.uppercase)
```
TIL about the `.textCase` modifier. Normally I would simply use `Text(category.label.uppercased)` to transform the string directly.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Add Product/ManualProductTypeOptions.swift:
```diff
@@ -6,37 +6,70 @@ struct ManualProductTypeOptions: View {
+                ForEach(productTypes, id: \.self) { productType in
+                    HStack(alignment: .top, spacing: Constants.margin) {
+                        Image(uiImage: productType.actionSheetImage.withRenderingMode(.alwaysTemplate))
+                            .font(.title3)
```
The modifier `font` will work only with system images. With custom assets, you might need to set the size explicitly with the`frame` modifier. To make it scalable, you'd need to get the scale from `ScaleMetric` property wrapper.

<img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/c6ecb122-0f55-4df4-93d8-2178ad9a88dc" width=320 />

As you see above, the custom icons are not scaled in large font sizes.

hafizrahman replied:
Updated in https://github.com/woocommerce/woocommerce-ios/pull/12273/commits/34442d134d84058e62685dcfb68a6ecdfc1bae34

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Add Product/ManualProductTypeOptions.swift:
```diff
@@ -6,37 +6,70 @@ struct ManualProductTypeOptions: View {
 private extension ManualProductTypeOptions {
     enum Constants {
+        // List of product categories. The ordering dictates how the categories are displayed.
+        static let productCategoriesOrder: [BottomSheetProductType.ProductCreationCategory] = [.standard, .subscription, .other]
```
Suggestion: You can omit this if you make `BottomSheetProductType.ProductCreationCategory` conform to `CaseIterable`. Then you can get the list of cases in the order declared in the enum with `BottomSheetProductType.ProductCreationCategory.allCases`.

hafizrahman replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/12273/commits/f1b37ed71271104480c29d1be2724a266c109d5b  also this comment is related to this https://github.com/woocommerce/woocommerce-ios/pull/12273#discussion_r1634539342

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/BottomSheetListSelector/BottomSheetProductType.swift:
```diff
@@ -199,6 +198,19 @@ public enum BottomSheetProductType: Hashable, Identifiable {
 
+    /// Product creation category (for UI purposes).
+    ///
+    var productCreationCategory: ProductCreationCategory {
+        switch self {
+        case .simple, .blank, .variable:
+            return .standard
+        case .subscription, .variableSubscription:
+            return .subscription
+        case .custom, .grouped, .affiliate:
+            return .other
+        }
+    }
```
Suggestion: you can avoid having to filter and match the category types if you replace this with a helper variable in `ProductCreationCategory` instead:

```Swift
extension ProductCreationCategory {
    var productTypes: BottomSheetProductType {
        // return the respective types for each category.
    }
}
```

Then in `ManualProductTypeOptions` you can get rid of `groupedProductTypes` and build the view by directly traversing the categories:

```Swift
ForEach(BottomSheetProductType.ProductCreationCategory.allCases, id: \.self) { category in
     ForEach(category.productTypes) { type in
          // TODO
     }
}
```

hafizrahman replied:
I did some refactor on https://github.com/woocommerce/woocommerce-ios/pull/12273/commits/f1b37ed71271104480c29d1be2724a266c109d5b based on your ideas here, but it doesn't fully implement your suggestion as I understood it.

The main thing is that the `ManualProductTypeOptions` receives `supportedProductTypes` that contains only the supported product types (calculated in  `presentActionSheetWithAI()` in `AddProductCoordinator`), so filtering things based on it is inevitable.

itsmeichigo reviewed (approved):
The update looks good! I wonder if we should also replace the bottom sheet for the case where AI is not eligible to use the same sheet but without the AI option. Currently, when testing for such sites, I still see the old sheet. Please feel free to file an issue for this as an enhancement separately though.

<img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/d18219e8-b177-443a-a101-c3ef7d9c6f7a" width=320 />

I left a few suggestions for code improvements below, but I'm pre-approving this PR.

rachelmcr commented:
@hafizrahman Is this PR being actively worked on? Just checking because it has been moved to a future milestone a handful of times already. (I'll go ahead and move it to the next milestone for now.)

hafizrahman commented:
@rachelmcr this was a hack week project that I didn't end up having enough time to complete, then got deprioritized over other tasks. Thanks for the reminder, I'll set as draft for now and come back to it next hack week.

itsmeichigo commented:
@hafizrahman It would be great to wrap this PR for this HACK week as well, it has been around for quite a while.

hafizrahman committed:
Merge branch 'trunk'

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Update image resizable and ForEach usage.

hafizrahman committed:
Refactor how product categorization and filtering is done to make it clearer.
